### DISCUSSION TITLE
Need a way to serialize simple objects

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#7435] [Scripting]
I'm trying to set up a multiplayer game and running into issues trying to send objects over the network.  At the moment, the RPC calls will only let you send primitives which makes trying to send a data structure with any complexity a real pain.  What would be helpful is to provide a way to serialize simple Resource objects to a PackedByteArray or something similar.  This could be easily sent across a network or saved to disk.  While you wouldn't be able to serialize Nodes, it should be possible to serialize objects that only have fields that are primitives or objects that are similarly simple.

I looked into ResourceSaver, but this only saves to disk, not of a PackedByteArray or String.  I also tried writing my own serializer using reflection, but I am unable to determine the script name of my object so I don't know what to create when it's time to deserialze the object (get_class() and get_script() are both returning null).

Anyhow, a serializer for simple objects would be really useful.

Edit:
I could give writing my own reflection based serializer another shot, but I really am having trouble trying to figure out what the data type of my object is through reflection.  Why is get_script() returning null?  I can see the script set in the inspector.  Some other forum posts recommend overriding get_class() with your script's class name, and that may have worked in Godot 3, but in Godot 4 it causes an error.  How do I get the script name of my object?

**Comment by theraot** [comment]
I have a binary serializer (which uses `StreamPeerBuffer`) for this kind of thing (which is also why I want `get_half` and `put_half`: https://github.com/godotengine/godot/pull/78872) *Addendum: It can handle object graphs, and can skip both serializing and deserializing specific types as a measure to avoid code injection.*

Serializing Nodes and built-in Resource types is possible, but I can tell that de-serializing some built-in types does not go smoothly, as there are properties that might result in an error when set depending on other properties. That is, the order in which you set them matters. At which point it is better to make a dedicated serializer for the particular type. Besides, often we want to serialize custom types anyway.

---

The method `get_class` is not virtual. You were not overriding it, you were hiding the base method, and that is now an error.

You can use `get_class` to get the built in class, and then use `get_script` for, well, the script.

I don't know why these methods would be giving you `null`. I can only imagine something else is going wrong that causes this. Can you make a minimum reproduction project? You might have an issue to report.

Assuming you got a `GDScript`, you can get its name by iterating over `ProjectSettings.get_global_class_list()` (I have plenty of places where I do). *I have become aware of `get_global_name`, but sadly that is not exposed to scripting. And the fact is that the answer to "If this enhancement will not be used often, can it be worked around with a few lines of script?" is yes. It is looping over `ProjectSettings.get_global_class_list()`.*

**Comment by blackears** [comment]
I just tried recreating what I was doing with get_script() in new code, and it seems to be working now.  I was doing was creating a new instance of my object and then calling get_script() on it sometime later.  I'm not sure what was going wrong.

**Reply by theraot** [reply]
I have an idea of what might be: `set("script", null)` will remove the script. So, for example, if you are resetting properties by iterating over the result of `get_property_list`, you might want to skip `"script"`.

Addendum: speaking of skipping. If the object is a `Resource`, I skip `"resource_local_to_scene"`, `"resource_name"` and `"resource_path"`.

**Comment by timothyqiu** [comment]
Have you tried `var_to_bytes()`?

**Reply by theraot** [reply]
If I'm not mistaken that does not handle objects, and OP is talking about serializing custom Resource objects. With that said, I suppose [`var_to_bytes_with_objects`](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#class-globalscope-method-var-to-bytes-with-objects) might do.

**Reply by timothyqiu** [reply]
@theraot Local objects can still be handled by `var_to_bytes()`. But yeah, it should be `var_to_bytes_with_objects()` if you want to send it outside.

**Reply by blackears** [reply]
I've tried var_to_bytes_with_objects, but it causes errors when you try to use it.  Apparently everything attached to the object is being serialized, including the script object itself.  So when you deserialize it tries to create a second definition of your class with the same name which causes an error to be thrown.

https://github.com/godotengine/godot/issues/80229#issuecomment-1665625739

In the end I just wrote my own serializer.  But I do feel this is something Godot should do.

**Reply by williamsokol** [reply]
There are definitely some major quirks with using the Godot serializing Commands: 
1. var_to_bytes(), 
2. var_to_bytes_with_objects, 
3. var_to_str()

But it is highlighted how to work around them in this issue:
https://github.com/godotengine/godot/issues/84956#issuecomment-3087304914
